File: lex-c/app-dev/src/LexAnalyzer.cpp
#include "LexAnalyzer.h"
#include "AppDocument.h"
#include <algorithm>
#include <iostream>

const int NOT_FOUND = 1000;

LexAnalyzer::LexAnalyzer(AppStateManager *stateManager) {
    this->stateManager = stateManager;
};

void LexAnalyzer::compile() {
    StringVector limiters = this->stateManager->getLimiters();
    StringVector spaces = this->stateManager->getSpaces();

    CompiledLineVector compiled;
    int i = 0;
    for (StringVector::iterator lim = limiters.begin(); lim != limiters.end(); ++lim) {
        compiled.push_back(
            CompiledLine({
                Table::LIMITERS,
                i,
                *lim
            })
        );
        i++;
    }

    i = 0;
    for (StringVector::iterator s = spaces.begin(); s != spaces.end(); ++s) {
        compiled.push_back(
            CompiledLine({
                Table::SPACES,
                i,
                *s
            })
        );
        i++;
    }
    this->stateManager->setCompiled(compiled);
}

void LexAnalyzer::parseText(std::string srcText) {
    std::string buf = srcText;

    int iter = 0;
    int MAX_ITER = this->stateManager->getStepNo();
    int currentPosInLine = 1;
    int lineNo = 1;

    AppDocument *document = new AppDocument();
    document->spaces = this->stateManager->getSpaces();
    document->limiters = this->stateManager->getLimiters();
    document->ids = StringVector({});
    document->compiled = this->stateManager->getCompiled();
    document->text = CanonicTextItemVector({});
    document->strings = StringVector({});

    while (buf.length() > 0 && iter < MAX_ITER) {
        IndexData indexData;
        this->getIndex(buf, document->compiled, &indexData);
        if (indexData.pos != NOT_FOUND) {
            bool limiterOrSpaceNotInBufStart = indexData.pos > 0;
            if (limiterOrSpaceNotInBufStart) {
                std::string newId = buf.substr(0, indexData.pos);
                this->addIdToText(newId, lineNo, currentPosInLine, document);
                currentPosInLine += newId.length();
                buf = buf.substr(indexData.pos);
            } else {
                this->addLimiterOrSpaceToText(
                    indexData,
                    lineNo,
                    currentPosInLine,
                    document
                );
                CompiledLine compiledLine = document->compiled[indexData.compiledLineIndex];
                currentPosInLine += compiledLine.lexem.length();
                buf = buf.substr(compiledLine.lexem.length());
                if (compiledLine.lexem == "\n") {
                    lineNo++;
                    currentPosInLine = 1;
                }
            }
        } else {
            this->addIdToText(buf, lineNo, currentPosInLine, document);
            buf = "";
        }
        iter++;
    }
    this->stateManager->setInputString(buf);
    this->stateManager->setText(document->text);
    this->stateManager->setIds(document->ids);
    this->stateManager->setLineNo(lineNo);
    this->stateManager->setCurrentPosInLine(currentPosInLine);
    delete document;
}

void LexAnalyzer::getIndex(std::string s, CompiledLineVector compiled, IndexData *result) {
    int bestIndex = NOT_FOUND;
    CompiledLine *bestCompiledLine = NULL;

    for (CompiledLineVector::iterator line = compiled.begin(); line != compiled.end(); ++line) {
        int pos = s.find(line->lexem.c_str());
        if (pos >= 0) {
            if (pos < bestIndex) {
                bestIndex = pos;
                bestCompiledLine = &*line;
            }
        }
    }
    result->pos = bestIndex;
    result->compiledLineIndex = bestCompiledLine != NULL ? bestCompiledLine - &*compiled.begin() : -1;
};

void LexAnalyzer::addIdToText (
    std::string newId,
    int lineNo,
    int currentPosInLine,
    AppDocument *doc
) {
    StringVector::iterator curId;
    for (curId = doc->ids.begin(); curId != doc->ids.end(); curId++ ) {
        if (*curId == newId) {
            break;
        }
    }
    CanonicTextItem newTextItem({
        Table::IDS,
        -1,
        lineNo,
        currentPosInLine,
        newId
    });

    if (curId == doc->ids.end()) {
        doc->ids.push_back(newId);
        newTextItem.tableIndex = doc->ids.size() - 1;
    } else {
        newTextItem.tableIndex = curId - doc->ids.begin();
    }
    doc->text.push_back(newTextItem);
};

void LexAnalyzer::addLimiterOrSpaceToText(
    IndexData indexData,
    int lineNo,
    int currentPosInLine,
    AppDocument *doc
) {
    CompiledLine compiledLine = doc->compiled[indexData.compiledLineIndex];
    CanonicTextItem newTextItem({
        compiledLine.tableId,
        compiledLine.tableIndex,
        lineNo,
        currentPosInLine,
        compiledLine.lexem
    });
    doc->text.push_back(newTextItem);
};

void LexAnalyzer::buildStrings() {
    CanonicTextItemVector srcText = this->stateManager->getText();
    std::string state = "work";
    CanonicTextItemVector buffer({});

    AppDocument *document = new AppDocument();
    document->spaces = this->stateManager->getSpaces();
    document->limiters = this->stateManager->getLimiters();
    document->ids = StringVector({});
    document->compiled = this->stateManager->getCompiled();
    document->text = CanonicTextItemVector({});
    document->strings = StringVector({});    

    CanonicTextItemVector::iterator token;
    for (token = srcText.begin(); token != srcText.end(); token++ ) {
        if (state == "work" && token->lexem == "'") {
            state = "openedString";
            continue;
        }
        if (state == "openedString" && token->lexem == "'") {
            std::string newId = this->concatString(buffer);

            CanonicTextItem token = buffer[0];
            this->addStringToText(newId, token.lineNo, token.pos, document);
            buffer = CanonicTextItemVector({});
            state = "work";
            continue;
        }
        if (state == "openedString" && token->lexem == "\n") {
            std::cout << "Незавершенная строковая константа" << std::endl;

            CanonicTextItemVector::iterator bufToken;
            for (bufToken = buffer.begin(); bufToken != buffer.end(); bufToken++ ) {
                document->text.push_back(*bufToken);
            };
        
            buffer = CanonicTextItemVector({});
            continue;
        }
        if (state == "openedString") {
            buffer.push_back(*token);
            continue;
        }
        if (state == "work") {
            if (token->tableId == Table::IDS) {
                this->addIdToText(token->lexem, token->lineNo, token->pos, document);
            } else {
                document->text.push_back(*token);
            }
            continue;
        }
    }    

    this->stateManager->setText(document->text);
    this->stateManager->setIds(document->ids);
    this->stateManager->setStrings(document->strings);

    delete document;
};

std::string LexAnalyzer::concatString (CanonicTextItemVector srcText) {
    std::string result;
    CanonicTextItemVector::iterator token;
    for (token = srcText.begin(); token != srcText.end(); token++ ) {
        result = result + token->lexem;
    }
    return result;
};

void LexAnalyzer::addStringToText (
    std::string newId,
    int lineNo,
    int currentPosInLine,
    AppDocument *doc
) {
    StringVector::iterator curString;
    for (curString = doc->strings.begin(); curString != doc->strings.end(); curString++ ) {
        if (*curString == newId) {
            break;
        }
    }
    CanonicTextItem newTextItem({
        Table::STRINGS,
        -1,
        lineNo,
        currentPosInLine,
        newId
    });

    if (curString == doc->strings.end()) {
        doc->strings.push_back(newId);
        newTextItem.tableIndex = doc->strings.size() - 1;
    } else {
        newTextItem.tableIndex = curString - doc->strings.begin();
    }
    doc->text.push_back(newTextItem);
};

void LexAnalyzer::removeComments() {
    CanonicTextItemVector srcText = this->stateManager->getText();
    std::string state = "work";
    CanonicTextItemVector buffer({});

    AppDocument *document = new AppDocument();
    document->spaces = this->stateManager->getSpaces();
    document->limiters = this->stateManager->getLimiters();
    document->ids = StringVector({});
    document->compiled = this->stateManager->getCompiled();
    document->text = CanonicTextItemVector({});
    document->strings = StringVector({});    

    CanonicTextItemVector::iterator token;
    for (token = srcText.begin(); token != srcText.end(); token++ ) {
        if (state == "work" && token->lexem == "/") {
            buffer.push_back(*token);
            state = "firstSlash";
            continue;
        }
        if (state == "firstSlash" && token->lexem == "/") {
            state = "comment";
            continue;
        }
        if (state == "comment" && token->lexem == "\n") {
            document->text.push_back(*token);
            state = "work";
            continue;
        }
        if (state == "firstSlash") {
            CanonicTextItem buf0 = buffer[0];
            document->text.push_back(buf0);
            buffer = CanonicTextItemVector({});
            document->text.push_back(*token);
            state = "work";
            continue;
        }
        if (state == "work") {
            switch (token->tableId) {
                case Table::IDS:
                    this->addIdToText(token->lexem, token->lineNo, token->pos, document);
                    break;

                case Table::STRINGS:
                    this->addStringToText(token->lexem, token->lineNo, token->pos, document);
                    break;

                default:
                    document->text.push_back(*token);
                    break;
            }
            continue;
        }

    };
    this->stateManager->setText(document->text);
    this->stateManager->setIds(document->ids);
    this->stateManager->setStrings(document->strings);
};

File: lex-holyjs-draft/app-dev/src/app/LexAnalyzer.ts
import {
    AppDocument,
    CanonicTextItem,
    CompiledLine,
    defaultCompiledLine,
    IndexData,
    Table
} from '@src/app.types';
import { AppStateManager } from '@src/AppStateManager';

const NOT_FOUND = 10000;

export class LexAnalyzer {
    constructor(private stateManager: AppStateManager) {}

    parseText = (srcText: string) => {
        let buf = srcText;

        let iter = 0;
        const MAX_ITER = this.stateManager.getStepNo();
        let currentPosInLine = 1;
        let lineNo = 1;

        let document: AppDocument = {
            spaces: this.stateManager.getSpaces(),
            limiters: this.stateManager.getLimiters(),
            ids: [],
            compiled: this.stateManager.getCompiled(),
            text: [],
            strings: []
        };

        while (buf.length > 0 && iter < MAX_ITER) {
            const indexData = this.getIndex(buf, document.compiled);
            if (indexData.pos !== NOT_FOUND) {
                const limiterOrSpaceNotInBufStart = indexData.pos > 0;
                if (limiterOrSpaceNotInBufStart) {
                    const newId = buf.substring(0, indexData.pos);
                    document = this.addIdToText(newId, lineNo, currentPosInLine, document);
                    currentPosInLine += newId.length;
                    buf = buf.substring(indexData.pos);
                } else {
                    document = this.addLimiterOrSpaceToText(
                        indexData,
                        lineNo,
                        currentPosInLine,
                        document
                    );
                    currentPosInLine += indexData.lexem.lexem.length;
                    buf = buf.substring(indexData.lexem.lexem.length);
                    if (indexData.lexem.lexem === '\n') {
                        lineNo++;
                        currentPosInLine = 1;
                    }
                }
            } else {
                document = this.addIdToText(buf, lineNo, currentPosInLine, document);
                buf = '';
            }
            iter++;
        }

        this.stateManager.setInputString(buf);
        this.stateManager.setText(document.text);
        this.stateManager.setIds(document.ids);
        this.stateManager.setLineNo(lineNo);
        this.stateManager.setCurrentPosInLine(currentPosInLine);
        return iter;
    };

    getIndex = (s: string, compiled: CompiledLine[]): IndexData => {
        let bestIndex = NOT_FOUND;
        let bestCompiledLine = defaultCompiledLine;
        compiled.forEach((line: CompiledLine) => {
            const index = s.indexOf(line.lexem);
            if (index >= 0) {
                if (index < bestIndex) {
                    bestIndex = index;
                    bestCompiledLine = line;
                }
            }
        });

        return { pos: bestIndex, lexem: bestCompiledLine };
    };

    buildStrings = () => {
        const srcText = this.stateManager.getText();
        let state = 'work';
        let buffer: CanonicTextItem[] = [];

        let document: AppDocument = {
            spaces: this.stateManager.getSpaces(),
            limiters: this.stateManager.getLimiters(),
            ids: [],
            compiled: this.stateManager.getCompiled(),
            text: [],
            strings: []
        };

        srcText.forEach((token: CanonicTextItem) => {
            if (state === 'work' && token.lexem === "'") {
                state = 'openedString';
                return;
            }
            if (state === 'openedString' && token.lexem === "'") {
                const newId = this.concatString(buffer);

                document = this.addStringToText(newId, buffer[0].lineNo, buffer[0].pos, document);
                buffer = [];
                state = 'work';
                return;
            }
            if (state === 'openedString' && token.lexem === '\n') {
                console.error('Незавершенная строковая константа', buffer[0].lineNo, buffer[0].pos);
                document.text = [...document.text, ...buffer];
                buffer = [];
                return;
            }

            if (state === 'openedString') {
                buffer = [...buffer, token];
                return;
            }

            if (state === 'work') {
                if (token.tableId === 'i') {
                    document = this.addIdToText(token.lexem, token.lineNo, token.pos, document);
                } else {
                    document.text = [...document.text, token];
                }
                return;
            }
        });

        this.stateManager.setText(document.text);
        this.stateManager.setIds(document.ids);
        this.stateManager.setStrings(document.strings);
    };

    addIdToText = (
        newId: string,
        lineNo: number,
        currentPosInLine: number,
        doc: AppDocument
    ): AppDocument => {
        const posInIds = doc.ids.findIndex((id: string) => id === newId);
        const newDoc = { ...doc };
        let newTextItem: CanonicTextItem = {
            tableId: Table.IDS,
            tableIndex: posInIds,
            lineNo,
            pos: currentPosInLine,
            lexem: newId
        };
        if (posInIds < 0) {
            newDoc.ids = [...newDoc.ids, newId];
            newTextItem.tableIndex = newDoc.ids.length - 1;
        }

        newDoc.text = [...newDoc.text, newTextItem];
        return newDoc;
    };

    concatString = (text: CanonicTextItem[]): string => {
        return text.map((token: CanonicTextItem) => token.lexem).join('');
    };

    addLimiterOrSpaceToText = (
        indexData: IndexData,
        lineNo: number,
        currentPosInLine: number,
        doc: AppDocument
    ) => {
        const newTextItem: CanonicTextItem = {
            tableId: indexData.lexem.tableId,
            tableIndex: indexData.lexem.tableIndex,
            lineNo,
            pos: currentPosInLine,
            lexem: indexData.lexem.lexem
        };
        return { ...doc, text: [...doc.text, newTextItem] };
    };

    addStringToText = (
        newId: string,
        lineNo: number,
        currentPosInLine: number,
        doc: AppDocument
    ): AppDocument => {
        const posInStrings = doc.strings.findIndex((id: string) => id === newId);
        const newDoc = { ...doc };
        let newTextItem: CanonicTextItem = {
            tableId: Table.STRINGS,
            tableIndex: posInStrings,
            lineNo,
            pos: currentPosInLine,
            lexem: newId
        };
        if (posInStrings < 0) {
            newDoc.strings = [...newDoc.strings, newId];
            newTextItem.tableIndex = newDoc.strings.length - 1;
        }

        newDoc.text = [...newDoc.text, newTextItem];
        return newDoc;
    };

    removeComments = () => {
        const srcText = this.stateManager.getText();
        let state = 'work';
        let buffer: CanonicTextItem[] = [];

        let document: AppDocument = {
            spaces: this.stateManager.getSpaces(),
            limiters: this.stateManager.getLimiters(),
            ids: [],
            compiled: this.stateManager.getCompiled(),
            text: [],
            strings: []
        };

        srcText.forEach((token: CanonicTextItem) => {
            if (state === 'work' && token.lexem === '/') {
                buffer = [token];
                state = 'firstSlash';
                return;
            }
            if (state === 'firstSlash' && token.lexem === '/') {
                state = 'comment';
                return;
            }
            if (state === 'comment' && token.lexem === '\n') {
                document.text = [...document.text, token];
                state = 'work';
                return;
            }

            if (state === 'firstSlash') {
                document.text = [...document.text, buffer[0]];
                buffer = [];
                document.text = [...document.text, token];
                state = 'work';
                return;
            }

            if (state === 'work') {
                switch (token.tableId) {
                    case Table.IDS:
                        document = this.addIdToText(token.lexem, token.lineNo, token.pos, document);
                        break;

                    case Table.STRINGS:
                        document = this.addStringToText(
                            token.lexem,
                            token.lineNo,
                            token.pos,
                            document
                        );
                        break;

                    default:
                        document.text = [...document.text, token];
                }
                return;
            }
        });

        this.stateManager.setText(document.text);
        this.stateManager.setIds(document.ids);
        this.stateManager.setStrings(document.strings);
    };

    removeWhitespace = () => {
        const srcText = this.stateManager.getText();
        let state = 'work';

        let document: AppDocument = {
            spaces: this.stateManager.getSpaces(),
            limiters: this.stateManager.getLimiters(),
            ids: [],
            compiled: this.stateManager.getCompiled(),
            text: [],
            strings: []
        };

        srcText.forEach((token: CanonicTextItem) => {
            if (state === 'work' && token.tableId === Table.SPACES) {
                state = 'work';
                return;
            }
            if (state === 'work') {
                switch (token.tableId) {
                    case Table.IDS:
                        document = this.addIdToText(token.lexem, token.lineNo, token.pos, document);
                        break;

                    case Table.STRINGS:
                        document = this.addStringToText(
                            token.lexem,
                            token.lineNo,
                            token.pos,
                            document
                        );
                        break;

                    default:
                        document.text = [...document.text, token];
                }
                return;
            }
        });

        this.stateManager.setText(document.text);
        this.stateManager.setIds(document.ids);
        this.stateManager.setStrings(document.strings);
    };

    compile = () => {
        const limiters = this.stateManager.getLimiters();
        const spaces = this.stateManager.getSpaces();

        let compiled: CompiledLine[] = [];
        compiled = [
            ...limiters.map(
                (lexem: string, index: number): CompiledLine => ({
                    tableId: Table.LIMITERS,
                    tableIndex: index,
                    lexem
                })
            ),
            ...spaces.map(
                (lexem: string, index: number): CompiledLine => ({
                    tableId: Table.SPACES,
                    tableIndex: index,
                    lexem
                })
            )
        ];
        this.stateManager.setCompiled(compiled);
        return this;
    };
}
